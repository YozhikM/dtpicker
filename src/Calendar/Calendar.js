/* @flow */
/* eslint-disable no-nested-ternary */

import React from 'react';
import {
  startOfMonth,
  endOfMonth,
  addMonths,
  setHours,
  setMinutes,
  setSeconds,
  format
} from 'date-fns';
import Button from '../Button/Button';
import CalendarDateTimePicker from './DateTimePicker/CalendarDateTimePicker';
import s from './DateTimePicker/MainCalendar.scss';
import { getValueFromDate, type Highlight } from './helpers';

export type StringValue = { min: string, max: string };

export type DateValue = { min: Date, max: Date };

type Value = { year: number, month: number, day?: number };

type Props = {
  value?: StringValue,
  highlight?: DateValue,
  minDate?: Date,
  maxDate?: Date,
  singleCalendar?: boolean,
  time?: boolean,
  hideResetBtn?: boolean,
  hideSubmitBtn?: boolean,
  onChange?: Function,
  onSubmit?: Function,
  toggleCalendar?: Function
};

type State = {
  highlight: Highlight | Highlight[],
  value: { min: Value, max: Value },
  singleCalendar: boolean,
  time: boolean,
  isCalendarShown: boolean,
  isNextClickAwaited: boolean,
  firstDate: Highlight,
  secondDate: Highlight
};

export default class Calendar extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const { firstDate, secondDate } = this.calculateRangeFromProps(props);
    const prevCurrentMonthValue = getValueFromDate(addMonths(new Date(), -1));
    const currentMonthValue = getValueFromDate();

    this.state = {
      value: props.value
        ? this.convertValueToStateValue(props.value)
        : props.highlight
          ? this.convertValueToStateValue(props.highlight)
          : {
              min: this.combineStateValue(prevCurrentMonthValue),
              max: this.combineStateValue(currentMonthValue)
            },
      highlight: props.highlight
        ? this.convertValueToArray(props.highlight)
        : props.value
          ? this.convertValueToArray(props.value)
          : this.getDefaultHighlight(props.singleCalendar),
      time: props.time || false,
      singleCalendar: props.singleCalendar || false,
      isCalendarShown: true,
      isNextClickAwaited: false,
      firstDate,
      secondDate
    };
  }

  combineStateValue = (value: any): Value => ({ ...value, day: value.day || 1 });

  getDefaultHighlight = (isSingleCalendar?: boolean): Highlight[] | Highlight => {
    const defaultHighlight = getValueFromDate();

    if (isSingleCalendar) {
      return defaultHighlight;
    }
    return [defaultHighlight, defaultHighlight];
  };

  calculateRangeFromProps = (props: Props): { firstDate: Highlight, secondDate: Highlight } => {
    const { minDate, maxDate } = this.props;
    const minRangeValue = getValueFromDate(minDate);
    const maxRangeValue = getValueFromDate(maxDate);

    if (props.value) {
      const { min, max } = props.value || {};
      const minDateValue = new Date(min);
      const maxDateValue = new Date(max);

      const maxDateRangeValue = getValueFromDate(maxDateValue);
      const minDateRangeValue = getValueFromDate(minDateValue);
      const minEndOfDayValue = minDate
        ? getValueFromDate(this.convertToEndOfDay(minDate))
        : getValueFromDate();
      const maxEndOfDayValue = maxDate
        ? getValueFromDate(this.convertToEndOfDay(maxDate))
        : getValueFromDate();

      if (minDate && !maxDate) {
        if (minDateValue < minDate && maxDateValue >= minDate) {
          return { firstDate: minRangeValue, secondDate: maxDateRangeValue };
        } else if (minDateValue < minDate && maxDateValue < minDate) {
          return { firstDate: minRangeValue, secondDate: minEndOfDayValue };
        }
      }

      if (maxDate) {
        if (maxDateValue > maxDate && minDateValue <= maxDate) {
          if (minDate && minDateValue >= minDate) {
            return { firstDate: minDateRangeValue, secondDate: maxEndOfDayValue };
          } else if (minDate && minDateValue < minDate) {
            return { firstDate: minRangeValue, secondDate: maxEndOfDayValue };
          } else if (!minDate) {
            return { firstDate: minDateRangeValue, secondDate: maxEndOfDayValue };
          }
        } else if (maxDateValue > maxDate && minDateValue > maxDate) {
          return { firstDate: maxRangeValue, secondDate: maxEndOfDayValue };
        }
      }

      return { firstDate: minDateRangeValue, secondDate: maxDateRangeValue };
    } else if (props.highlight) {
      const { min, max } = props.highlight || {};
      const minRangeHighlight = getValueFromDate(min);
      const maxRangeHighlight = getValueFromDate(max);

      if (minDate && min < minDate) {
        return { firstDate: minRangeValue, secondDate: maxRangeHighlight };
      } else if (maxDate && max > maxDate) {
        return { firstDate: minRangeHighlight, secondDate: maxRangeValue };
      }

      return { firstDate: minRangeHighlight, secondDate: maxRangeHighlight };
    }
    return { firstDate: getValueFromDate(), secondDate: getValueFromDate() };
  };

  convertValueToStateValue = (value: StringValue | DateValue): { min: Value, max: Value } => {
    const { minDate, maxDate } = this.props;
    const { min, max } = value || {};

    const minDateValue = this.convertToStartOfDay(new Date(min));
    const maxDateValue = this.convertToEndOfDay(new Date(max));
    const nextMonth: Date = startOfMonth(addMonths(minDateValue, 1));
    const prevMonth: Date = endOfMonth(addMonths(minDateValue, -1));

    const prevMonthValue = getValueFromDate(prevMonth);
    const nextMonthValue = getValueFromDate(nextMonth);
    const minValue = getValueFromDate(minDateValue);
    const maxValue = getValueFromDate(maxDateValue);
    const prevMonthMaxValue = getValueFromDate(addMonths(maxDate, -1));
    const nextMonthMaxValue = getValueFromDate(addMonths(minDate, 1));

    if (maxDate && nextMonth > maxDate) {
      return {
        min: this.combineStateValue(prevMonthValue),
        max: this.combineStateValue(minValue)
      };
    }

    if (maxDate) {
      if (maxDateValue > maxDate && nextMonth > maxDate) {
        return {
          min: this.combineStateValue(prevMonthMaxValue),
          max: this.combineStateValue(maxValue)
        };
      }
    }

    if (minDate) {
      if (minDateValue < minDate && prevMonth < minDate) {
        return {
          min: this.combineStateValue(minValue),
          max: this.combineStateValue(nextMonthMaxValue)
        };
      }
    }

    return {
      min: this.combineStateValue(minValue),
      max: this.combineStateValue(nextMonthValue)
    };
  };

  convertValueToArray = (value: StringValue | DateValue): Highlight[] | Highlight => {
    const { minDate, maxDate } = this.props;
    const arrayOfDate = Object.keys(value).map(key => {
      const minValue = new Date(value.min);
      const maxValue = new Date(value.max);

      if (minDate) {
        if (minValue < minDate) return getValueFromDate(minDate);
      } else if (maxDate) {
        if (maxValue > maxDate) return getValueFromDate(maxDate);
      }
      const date = new Date(value[key]);
      return getValueFromDate(date);
    });

    if (this.props.singleCalendar) return arrayOfDate[0];
    return arrayOfDate;
  };

  convertToStartOfDay = (date: Date): Date => setSeconds(setMinutes(setHours(date, 0), 0), 0);

  convertToEndOfDay = (date: Date): Date => setSeconds(setMinutes(setHours(date, 23), 59), 59);

  getValue = (first: Highlight, second: Highlight): StringValue => {
    let value;
    const firstDate = new Date(Date.UTC(first.year, first.month, first.day));
    const secondDate = new Date(Date.UTC(second.year, second.month, second.day));
    if (firstDate < secondDate) {
      value = {
        min: format(firstDate),
        max: format(this.convertToEndOfDay(secondDate))
      };
    } else {
      value = {
        min: format(secondDate),
        max: format(this.convertToEndOfDay(firstDate))
      };
    }
    return value;
  };

  onSetFirstDate = (date: Value): void => {
    const { onChange } = this.props;
    if (!Array.isArray(this.state.highlight)) return;

    const highlight = this.state.highlight.slice();

    if (highlight.length > 1) {
      highlight.pop();
      highlight.push({ ...date, day: date.day || 1, hour: 0, minute: 0, second: 0 });
    }

    const [secondDate, firstDate] = highlight;
    this.setState(
      {
        firstDate,
        secondDate,
        highlight
      },
      () => {
        if (onChange) {
          const value = this.getValue(firstDate, secondDate);
          onChange({}, value);
        }
      }
    );
  };

  onSetSecondDate = (date: Value): void => {
    const { onChange } = this.props;
    if (!Array.isArray(this.state.highlight)) return;

    const highlight = this.state.highlight.slice();

    if (highlight.length > 1) {
      highlight.shift();
      highlight.unshift({ ...date, day: date.day || 1, hour: 0, minute: 0, second: 0 });
    }

    const [secondDate, firstDate] = highlight;
    this.setState(
      {
        firstDate,
        secondDate,
        highlight
      },
      () => {
        if (onChange) {
          const value = this.getValue(firstDate, secondDate);
          onChange({}, value);
        }
      }
    );
  };

  onSetDate = (date: Value): void => {
    const { onChange } = this.props;
    const highlight = { ...date, day: date.day || 1, hour: 0, minute: 0, second: 0 };

    this.setState({ highlight }, () => {
      if (onChange) {
        const value = this.getValue(highlight, highlight);
        onChange({}, value);
      }
    });
  };

  onSetDateByClick = (date: Value): void => {
    const { onChange } = this.props;
    const { isNextClickAwaited } = this.state;
    if (!Array.isArray(this.state.highlight)) return;

    const highlight = this.state.highlight.slice();

    if (highlight.length > 1) {
      highlight.shift();
      highlight.push({ ...date, day: date.day || 1, hour: 0, minute: 0, second: 0 });
    }

    if (!isNextClickAwaited) {
      const [secondDate, firstDate] = highlight;
      this.setState(
        {
          isNextClickAwaited: true,
          firstDate,
          secondDate,
          highlight
        },
        () => {
          if (onChange) {
            const value = this.getValue(firstDate, secondDate);
            onChange({}, value);
          }
        }
      );
    }

    if (isNextClickAwaited) {
      const [secondDate, firstDate] = highlight;
      this.setState(
        {
          isNextClickAwaited: false,
          firstDate,
          secondDate,
          highlight
        },
        () => {
          if (onChange) {
            const value = this.getValue(firstDate, secondDate);
            onChange({}, value);
          }
        }
      );
    }
  };

  onChangeMin = ({ year, month, day = 1 }: Value): void => {
    const cloneValue = {};
    cloneValue.min = { year, month, day };
    cloneValue.max = { year, month: month + 1, day };

    this.setState({ value: cloneValue });
  };

  onChangeMax = ({ year, month, day = 1 }: Value): void => {
    const cloneValue = {};
    cloneValue.max = { year, month, day };
    cloneValue.min = { year, month: month - 1, day };

    this.setState({ value: cloneValue });
  };

  onChangeSingleCalendar = ({ year, month, day }: Value): void => {
    const cloneValue = {};
    const date = { year, month, day };
    cloneValue.min = date;
    cloneValue.max = date;

    this.setState({ value: cloneValue });
  };

  onChangeCalendarVisibility = (): void => {
    const { isCalendarShown } = this.state;
    this.setState({ isCalendarShown: !isCalendarShown });
  };

  resetState = () => {
    const dateNow = getValueFromDate(new Date());
    this.setState({
      firstDate: dateNow,
      secondDate: dateNow,
      highlight: [dateNow, dateNow]
    });
  };

  onSubmit = () => {
    const { highlight } = this.state;
    const { onSubmit, toggleCalendar } = this.props;
    let stringValue;
    if (Array.isArray(highlight)) {
      const [firstDate, secondDate] = highlight;
      stringValue = this.getValue(firstDate, secondDate);
    } else {
      stringValue = this.getValue(highlight, highlight);
    }

    if (onSubmit) onSubmit({}, stringValue);
    if (toggleCalendar) toggleCalendar();
  };

  renderSingleCalendar = () => {
    const { maxDate, minDate } = this.props;
    const { time, value, highlight, isCalendarShown } = this.state;
    const { min } = value || {};
    if (Array.isArray(highlight)) return null;

    return (
      <div className={s.container}>
        <CalendarDateTimePicker
          time={time}
          value={{ year: min.year, month: min.month, day: min.day }}
          highlight={highlight}
          visibleDate={highlight}
          isCalendarShown={isCalendarShown}
          onSetDate={this.onSetDate}
          onChange={this.onChangeSingleCalendar}
          onChangeCalendarVisibility={this.onChangeCalendarVisibility}
          maxDate={maxDate}
          minDate={minDate}
          leftArrow
          rightArrow
        />
      </div>
    );
  };

  renderDualCalendar = () => {
    const { maxDate, minDate, hideResetBtn, hideSubmitBtn } = this.props;
    const { time, value, highlight, isCalendarShown, firstDate, secondDate } = this.state;
    const { min, max } = value || {};

    return (
      <div>
        <div className={s.container}>
          <CalendarDateTimePicker
            time={time}
            value={{ year: min.year, month: min.month, day: min.day }}
            highlight={highlight}
            visibleDate={firstDate}
            onChange={this.onChangeMin}
            onSetDate={this.onSetFirstDate}
            isCalendarShown={isCalendarShown}
            onSetDateByClick={this.onSetDateByClick}
            onChangeCalendarVisibility={this.onChangeCalendarVisibility}
            maxDate={maxDate}
            minDate={minDate}
            leftArrow
          />
          <CalendarDateTimePicker
            time={time}
            value={{ year: max.year, month: max.month, day: max.day }}
            highlight={highlight}
            visibleDate={secondDate}
            onChange={this.onChangeMax}
            onSetDate={this.onSetSecondDate}
            isCalendarShown={isCalendarShown}
            onSetDateByClick={this.onSetDateByClick}
            onChangeCalendarVisibility={this.onChangeCalendarVisibility}
            maxDate={maxDate}
            minDate={minDate}
            rightArrow
          />
        </div>
        <div className={s.calendarBtn}>
          {!hideResetBtn && (
            <Button onClick={this.resetState} red xs>
              Очистить
            </Button>
          )}
          {!hideSubmitBtn && (
            <Button onClick={this.onSubmit} blue filled xs>
              Подтвердить
            </Button>
          )}
        </div>
      </div>
    );
  };

  render() {
    const { highlight, singleCalendar } = this.state;

    if (singleCalendar && !Array.isArray(highlight)) {
      return this.renderSingleCalendar();
    }

    return this.renderDualCalendar();
  }
}
