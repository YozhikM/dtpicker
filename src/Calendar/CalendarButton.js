/* @flow */

import * as React from 'react';
import SvgIcon from '../SvgIcon/SvgIcon';

type Props = {
  onClick?: Function,
  style?: Object,
};

export default class CalendarButton extends React.Component<Props, void> {
  render() {
    const { onClick, style } = this.props;
    return (
      <div onClick={onClick} style={style}>
        <SvgIcon file="calendar" />
      </div>
    );
  }
}
