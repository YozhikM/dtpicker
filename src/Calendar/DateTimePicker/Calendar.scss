.btns {
  display: flex;
}

.icon {
  display: flex;
  width: 24px;
  height: 24px;
  align-items: center;
  justify-content: center;
  position: absolute;
  right: 24px;
  top: 5px;
  cursor: pointer;
}

.root {
  position: relative;
  bottom: 28px;

  table {
    border-collapse: collapse;
  }

  tr {
    display: flex;
  }

  tr:nth-of-type(1) {
    justify-content: flex-end;
  }

  td,
  th {
    width: 30px;
    height: 30px;
    text-align: center;
    line-height: 30px;
    padding: 0;
    box-sizing: border-box;
    cursor: pointer;
    outline: none;
  }
}

.button_container {
  display: flex;
  justify-content: space-around;

  button {
    display: block;
    text-align: center;
    margin: 3% 0 3% 0;
    color: #3c3f40;
    font-size: 14px;
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
  }
}

.button_single_container {
  position: absolute;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 45px;
  z-index: -1;

  button {
    display: block;
    text-align: center;
    margin: 0;
    color: #3c3f40;
    font-size: 14px;
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
    transform: scale(1);
    transition: transform 0.6s ease;
  }
  button:hover {
    transform: scale(0.8);
    transition: transform 0.6s ease;
  }
}
.left_arrow {
  align-self: flex-start;
  z-index: 2;
  color: #333 !important;
  margin: 0 !important;
  border: none !important;
  background-color: transparent !important;
  padding: 0 !important;
  position: relative;
  top: 3px;
}
.right_arrow {
  align-self: flex-end;
  z-index: 2;
  color: #333 !important;
  margin: 0 !important;
  border: none !important;
  background-color: transparent !important;
  padding: 0 !important;
  position: relative;
  top: 3px;
}
.calendar_container {
  .span {
    margin: 3% auto;
    font-size: 18px;
    font-weight: bold;
    cursor: pointer;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }

  button {
    display: block;
    margin: 3% auto;
  }
}
.table {
  display: inline-block;
  z-index: 2;
}

.c_button_container {
  display: flex;
  justify-content: space-between;

  button {
    display: block;
    text-align: center;
    margin: 3% 0 3% 0;
    color: #3c3f40;
    font-size: 14px;
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
    z-index: 1;
  }
}

.time_picker {
  text-align: center;

  button {
    cursor: pointer;
    border: none;
    background-color: transparent;
    outline: none;
    margin: 0 auto;
    padding: 0 !important;
  }
}
.container {
  display: flex;
  align-items: center;
  margin: 0 2px;
  flex-direction: column;
}

.arrow {
  position: absolute;
  top: 5px;
  z-index: 3;
}

.control_panel {
  position: absolute;
  bottom: 50%;
  display: flex;
  flex-flow: row wrap;
  align-items: center;
  justify-content: space-around;
  width: 150px;
  left: 10%;
  z-index: 3;

  button {
    cursor: pointer;
    background: transparent;
    border-radius: 50px;
    border: 1px solid rgb(52, 73, 94);
    padding: 0;
    width: 36px;
    height: 36px;
    display: flex;
    justify-content: center;
    align-items: center;
    outline: none;
  }
}
