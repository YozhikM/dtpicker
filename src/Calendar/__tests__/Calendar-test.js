/* @flow */

import * as React from 'react';
import Calendar from '../index';

describe('Calendar', () => {
  const stringValue = { min: '2010-03-02T18:00:00.000Z', max: '2010-03-02T17:59:59.000Z' };
  describe('render()', () => {
    it('basic', () => {
      const calendar = shallow(<Calendar value={stringValue} />);
      expect(calendar).toMatchSnapshot();
    });

    it('if singleCalendar', () => {
      const singleCalendar = shallow(<Calendar value={stringValue} singleCalendar />);

      expect(singleCalendar).toMatchSnapshot();
    });

    it('if time', () => {
      const timeCalendar = shallow(<Calendar value={stringValue} time />);

      expect(timeCalendar).toMatchSnapshot();
    });

    it('if time', () => {
      const wideCalendar = shallow(<Calendar value={stringValue} wide />);

      expect(wideCalendar).toMatchSnapshot();
    });
  });

  it('convertValueToDate()', () => {
    const wrapper = shallow(<Calendar value={stringValue} />);
    const convertedValue = wrapper.instance().convertValueToDate(stringValue);

    expect(convertedValue).toMatchSnapshot();
  });

  describe('convertValueToArray()', () => {
    it('if !singleCalendar', () => {
      const wrapper = shallow(<Calendar value={stringValue} />);
      const convertedValue = wrapper.instance().convertValueToArray(stringValue);

      expect(convertedValue).toMatchSnapshot();
    });

    it('if singleCalendar', () => {
      const wrapper = shallow(<Calendar singleCalendar value={stringValue} />);
      const convertedValue = wrapper.instance().convertValueToArray(stringValue);

      expect(convertedValue).toMatchSnapshot();
    });
  });

  it('convertToEndOfDay()', () => {
    const wrapper = shallow(<Calendar value={stringValue} />);
    const convertedDate = wrapper.instance().convertToEndOfDay(new Date(2010, 2, 3));

    expect(convertedDate).toEqual(new Date('2010-03-03T17:59:59.000Z'));
  });

  describe('getValue()', () => {
    it('if firstDate < secondDate', () => {
      const wrapper = shallow(<Calendar value={stringValue} />);
      const value = wrapper.instance().getValue(new Date(2010, 2, 3), new Date(2010, 2, 4));

      expect(value).toMatchSnapshot();
    });

    it('if firstDate > secondDate', () => {
      const wrapper = shallow(<Calendar value={stringValue} />);
      const value = wrapper.instance().getValue(new Date(2010, 2, 26), new Date(2010, 2, 1));

      expect(value).toMatchSnapshot();
    });
  });

  it('onSetFirstDate()', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Calendar value={stringValue} onChange={onChange} />);
    wrapper
      .find('CalendarDateTimePicker')
      .at(0)
      .props()
      .onSetDate(new Date(1994, 7, 1));

    const highlight = [new Date('2010-03-02T18:00:00.000Z'), new Date('1994-07-31T17:00:00.000Z')];
    const value = { max: '2010-03-03T17:59:59.000Z', min: '1994-07-31T17:00:00.000Z' };

    expect(wrapper.state('firstDate')).toEqual(new Date('1994-07-31T17:00:00.000Z'));
    expect(wrapper.state('secondDate')).toEqual(new Date('2010-03-02T18:00:00.000Z'));
    expect(wrapper.state('highlight')).toEqual(highlight);
    expect(onChange).toBeCalledWith({}, value);
  });

  it('onSetSecondDate()', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Calendar value={stringValue} onChange={onChange} />);
    wrapper
      .find('CalendarDateTimePicker')
      .at(1)
      .props()
      .onSetDate(new Date(2017, 12, 1));

    const highlight = [new Date('2017-12-31T18:00:00.000Z'), new Date('2010-03-02T17:59:59.000Z')];
    const value = { max: '2018-01-01T17:59:59.000Z', min: '2010-03-02T17:59:59.000Z' };

    expect(wrapper.state('firstDate')).toEqual(new Date('2010-03-02T17:59:59.000Z'));
    expect(wrapper.state('secondDate')).toEqual(new Date('2017-12-31T18:00:00.000Z'));
    expect(wrapper.state('highlight')).toEqual(highlight);
    expect(onChange).toBeCalledWith({}, value);
  });

  it('onSetDate()', () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Calendar value={stringValue} onChange={onChange} singleCalendar />);
    wrapper
      .find('CalendarDateTimePicker')
      .props()
      .onSetDate(new Date(2017, 12, 1));

    const highlight = new Date('2017-12-31T18:00:00.000Z');
    const value = { max: '2018-01-01T17:59:59.000Z', min: '2017-12-31T18:00:00.000Z' };

    expect(wrapper.state('firstDate')).toEqual(new Date('2010-03-02T18:00:00.000Z'));
    expect(wrapper.state('secondDate')).toEqual(new Date('2010-03-02T17:59:59.000Z'));
    expect(wrapper.state('highlight')).toEqual(highlight);
    expect(onChange).toBeCalledWith({}, value);
  });

  describe('onSetDateByClick()', () => {
    it('if !isNextClickAwaited', () => {
      const onChange = jest.fn();
      const wrapper = shallow(<Calendar value={stringValue} onChange={onChange} />);
      wrapper
        .find('CalendarDateTimePicker')
        .at(0)
        .props()
        .onSetDateByClick(new Date(2011, 12, 1));

      const highlight = [
        new Date('2010-03-02T17:59:59.000Z'),
        new Date('2011-12-31T18:00:00.000Z'),
      ];
      const value = { max: '2012-01-01T17:59:59.000Z', min: '2010-03-02T17:59:59.000Z' };

      expect(wrapper.state('isNextClickAwaited')).toBeTruthy();
      expect(wrapper.state('firstDate')).toEqual(new Date('2011-12-31T18:00:00.000Z'));
      expect(wrapper.state('secondDate')).toEqual(new Date('2010-03-02T17:59:59.000Z'));
      expect(wrapper.state('highlight')).toEqual(highlight);
      expect(onChange).toBeCalledWith({}, value);
    });

    it('if isNextClickAwaited', () => {
      const onChange = jest.fn();
      const wrapper = shallow(<Calendar value={stringValue} onChange={onChange} />);
      wrapper.setState({ isNextClickAwaited: true });
      wrapper.update();
      wrapper
        .find('CalendarDateTimePicker')
        .at(0)
        .props()
        .onSetDateByClick(new Date(2009, 1, 19));

      const highlight = [
        new Date('2010-03-02T17:59:59.000Z'),
        new Date('2009-02-18T18:00:00.000Z'),
      ];
      const value = { max: '2010-03-02T17:59:59.000Z', min: '2009-02-18T18:00:00.000Z' };

      expect(wrapper.state('isNextClickAwaited')).toBeFalsy();
      expect(wrapper.state('firstDate')).toEqual(new Date('2009-02-18T18:00:00.000Z'));
      expect(wrapper.state('secondDate')).toEqual(new Date('2010-03-02T17:59:59.000Z'));
      expect(wrapper.state('highlight')).toEqual(highlight);
      expect(onChange).toBeCalledWith({}, value);
    });
  });

  it('onChangeMin()', () => {
    const wrapper = shallow(<Calendar value={stringValue} />);
    wrapper
      .find('CalendarDateTimePicker')
      .at(0)
      .props()
      .onChange(new Date(2019, 4, 23));

    const value = {
      max: new Date('2019-06-22T18:00:00.000Z'),
      min: new Date('2019-05-22T18:00:00.000Z'),
    };

    expect(wrapper.state('value')).toEqual(value);
  });

  it('onChangeMax()', () => {
    const wrapper = shallow(<Calendar value={stringValue} />);
    wrapper
      .find('CalendarDateTimePicker')
      .at(1)
      .props()
      .onChange(new Date(2009, 9, 23));

    const value = {
      max: new Date('2009-10-22T18:00:00.000Z'),
      min: new Date('2009-09-22T18:00:00.000Z'),
    };

    expect(wrapper.state('value')).toEqual(value);
  });

  it('onChangeSingleCalendar()', () => {
    const wrapper = shallow(<Calendar singleCalendar value={stringValue} />);
    wrapper
      .find('CalendarDateTimePicker')
      .props()
      .onChange(new Date(2018, 9, 23));

    const value = {
      max: new Date('2018-10-22T18:00:00.000Z'),
      min: new Date('2018-10-22T18:00:00.000Z'),
    };

    expect(wrapper.state('value')).toEqual(value);
  });

  it('onChangeCalendarVisibility()', () => {
    const wrapper = shallow(<Calendar singleCalendar value={stringValue} />);
    wrapper
      .find('CalendarDateTimePicker')
      .props()
      .onChangeCalendarVisibility();

    expect(wrapper.state('isCalendarShown')).toBeFalsy();
  });

  it('renderSingleCalendar()', () => {
    const wrapper = shallow(<Calendar singleCalendar value={stringValue} />);
    const render = wrapper.instance().renderSingleCalendar();

    expect(render).toMatchSnapshot();
  });

  it('renderSingleCalendar()', () => {
    const wrapper = shallow(<Calendar value={stringValue} />);
    const render = wrapper.instance().renderDualCalendar();

    expect(render).toMatchSnapshot();
  });
});
