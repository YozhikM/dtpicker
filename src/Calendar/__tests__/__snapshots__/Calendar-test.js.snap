// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar convertValueToArray() if !singleCalendar 1`] = `
Array [
  2010-03-02T18:00:00.000Z,
  2010-03-02T17:59:59.000Z,
]
`;

exports[`Calendar convertValueToArray() if singleCalendar 1`] = `2010-03-02T18:00:00.000Z`;

exports[`Calendar convertValueToDate() 1`] = `
Object {
  "max": 2010-04-02T18:00:00.000Z,
  "min": 2010-03-02T18:00:00.000Z,
}
`;

exports[`Calendar getValue() if firstDate < secondDate 1`] = `
Object {
  "max": "2010-03-04T17:59:59.000Z",
  "min": "2010-03-02T18:00:00.000Z",
}
`;

exports[`Calendar getValue() if firstDate > secondDate 1`] = `
Object {
  "max": "2010-03-26T17:59:59.000Z",
  "min": "2010-02-28T18:00:00.000Z",
}
`;

exports[`Calendar render() basic 1`] = `
<div>
  <div
    className="undefined"
  >
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      leftArrow={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      time={false}
      value={2010-03-02T18:00:00.000Z}
      visibleDate={2010-03-02T18:00:00.000Z}
    />
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      rightArrow={true}
      time={false}
      value={2010-04-02T18:00:00.000Z}
      visibleDate={2010-03-02T17:59:59.000Z}
    />
  </div>
</div>
`;

exports[`Calendar render() if singleCalendar 1`] = `
<div>
  <CalendarDateTimePicker
    highlight={2010-03-02T18:00:00.000Z}
    isCalendarShown={true}
    leftArrow={true}
    onChange={[Function]}
    onChangeCalendarVisibility={[Function]}
    onSetDate={[Function]}
    rightArrow={true}
    time={false}
    value={2010-03-02T18:00:00.000Z}
    visibleDate={2010-03-02T18:00:00.000Z}
  />
</div>
`;

exports[`Calendar render() if time 1`] = `
<div>
  <div
    className="undefined"
  >
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      leftArrow={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      time={true}
      value={2010-03-02T18:00:00.000Z}
      visibleDate={2010-03-02T18:00:00.000Z}
    />
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      rightArrow={true}
      time={true}
      value={2010-04-02T18:00:00.000Z}
      visibleDate={2010-03-02T17:59:59.000Z}
    />
  </div>
</div>
`;

exports[`Calendar render() if time 2`] = `
<div>
  <div
    className="undefined"
  >
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      leftArrow={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      time={false}
      value={2010-03-02T18:00:00.000Z}
      visibleDate={2010-03-02T18:00:00.000Z}
    />
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      rightArrow={true}
      time={false}
      value={2010-04-02T18:00:00.000Z}
      visibleDate={2010-03-02T17:59:59.000Z}
    />
  </div>
</div>
`;

exports[`Calendar renderSingleCalendar() 1`] = `
<div
  className={undefined}
>
  <CalendarDateTimePicker
    highlight={2010-03-02T18:00:00.000Z}
    isCalendarShown={true}
    leftArrow={true}
    onChange={[Function]}
    onChangeCalendarVisibility={[Function]}
    onSetDate={[Function]}
    rightArrow={true}
    time={false}
    value={2010-03-02T18:00:00.000Z}
    visibleDate={2010-03-02T18:00:00.000Z}
  />
</div>
`;

exports[`Calendar renderSingleCalendar() 2`] = `
<div>
  <div
    className="undefined"
  >
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      leftArrow={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      time={false}
      value={2010-03-02T18:00:00.000Z}
      visibleDate={2010-03-02T18:00:00.000Z}
    />
    <CalendarDateTimePicker
      highlight={
        Array [
          2010-03-02T18:00:00.000Z,
          2010-03-02T17:59:59.000Z,
        ]
      }
      isCalendarShown={true}
      onChange={[Function]}
      onChangeCalendarVisibility={[Function]}
      onSetDate={[Function]}
      onSetDateByClick={[Function]}
      rightArrow={true}
      time={false}
      value={2010-04-02T18:00:00.000Z}
      visibleDate={2010-03-02T17:59:59.000Z}
    />
  </div>
</div>
`;
